### PR TITLE
Verify sns message signature

### PR DESCRIPTION
**Pull request by WtfJoke**
Fixes gh-176

## :loudspeaker: Type of change

- [x] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring

## :scroll: Description

Verify the signature of a SNS Message by utilizing [SnsMessageManager#parseMessage](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/sns/message/SnsMessageManager.html#parseMessage-java.io.InputStream-) which also validates the Signature using SignatureVerifier. The downside of it, it does more than just validate the Signature, but I think it doesnt hurt.

As an alternative we could also validate the Signature using [SignatureChecker (aws sdk v1)](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/sns/util/SignatureChecker.html), but the class is deprecated and they suggest using SnsMessageManager.

In aws sdk 2 there would be a more optimal approach using [SignatureChecker (aws sdk v2)](https://github.com/aws/aws-sdk-java-v2/blob/master/services/sns/src/it/java/software/amazon/awssdk/services/sns/SignatureChecker.java#L40), but we still have aws sdk 1 here.

## :bulb: Motivation and Context

#176

## :green_heart: How did you test it?
I'll added two junit tests, which tests two cases (invalid/valid signature). 
For that I needed to add an additional constructor in order to be able to mock SnsMessageManager. I hope thats ok (appreciate comments for better ideas :)). 

I also verified it, by publishing some messages in my sns topic through the aws management console.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [N/A] I updated reference documentation to reflect the change
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

**Review by maciejwalkowiak — APPROVED** [review]
Thanks! We will include it in 2.4

**Review by MatejNedic — CHANGES_REQUESTED** [review]
@WtfJoke thanks for contribution!
I have an issue that we should address to have PR ready for Merge

**Comment by WtfJoke** [comment]
Just wanted to let you know that I will soon take care of the requested changes.

File: spring-cloud-aws-messaging/src/main/java/io/awspring/cloud/messaging/support/converter/NotificationRequestConverter.java [hunk 1]
```diff
@@ -46,8 +51,15 @@
 
 	private final MessageConverter payloadConverter;
 
+	private final SnsMessageManager snsMessageManager;
```
[the remark this answers is not in the document]
Comment (WtfJoke) [reply spring-cloud-aws-messaging/src/main/java/io/awspring/cloud/messaging/support/converter/NotificationRequestConverter.java:54]: @maciejwalkowiak First of nice find.

I know its been a while, but I dont understand the recommended approach fully 😅.  `SnsMessageManager` has the default constructor (which uses the wrong region) and the one where I pass the region. So I should use that one. 

How and where can I inject the correct region? I suppose I shouldnt duplicate the code in the `SnsAutoConfiguration`, right? Or should I simply inject `SNSProperties` and use this region (but which region provider do I use if the property is null, the default one)?

[the remark this answers is not in the document]
Comment (MatejNedic) [reply spring-cloud-aws-messaging/src/main/java/io/awspring/cloud/messaging/support/converter/NotificationRequestConverter.java:54]: Personally, I would create `SnsMessageManager` bean in `SnsAutoConfiguration` and pass it to `NotificationRequestConverter`.
For a region use `regionProvider` located in `SnsAutoConfiguration`.

@maciejwalkowiak @eddumelendez WDYT? Could you assist further?

[the remark this answers is not in the document]
Comment (WtfJoke) [reply spring-cloud-aws-messaging/src/main/java/io/awspring/cloud/messaging/support/converter/NotificationRequestConverter.java:54]: Your explaination helped, thank you. I pushed the changes. 

I always struggle a bit with formatting issues 😅 I've only managed to do this on my linux machine.
Havent yet tested the changes, but I think you still can take a look at it.

[the remark this answers is not in the document]
Comment (WtfJoke) [reply spring-cloud-aws-messaging/src/main/java/io/awspring/cloud/messaging/support/converter/NotificationRequestConverter.java:54]: Strangely enough, all tests are green on my machine :S

**Review by maciejwalkowiak — CHANGES_REQUESTED** [review]
I've had another look and I think we approached in from a wrong side. 

Signature verification should happen for messages sent to an HTTP endpoint, not ones sent to SQS queue. This means, we do not need to verify signature in `QueueMessageHandler` but rather somewhere in `NotificationMessageHandlerMethodArgumentResolver`. SQS related code should ideally not need to change. 

I know it's been a while, so is this something @WtfJoke you would be still interested in doing?

**Comment by WtfJoke** [comment]
@maciejwalkowiak no worries, I know the struggle of the life of a maintainer :D

However currently im quite occupied, so I dont want to block anybody which is interested in doing that.  Feel free to close the PR.

Thank you both for taking a look to the PR in the past.

**Pull request closed without merging**